src/pre_processing/extract_landscape.py

```diff
@@ -0,0 +1,165 @@
+
+    A list of numpy arrays containing position of specific positions of a face. if no face found a [62,2]
+    zero vector is placed in matrix
+
+    Keyword Arguments:
+        specific_part -- can specify to return vectors for specified facial features:
+                         [RIGHT_BROW, LEFT_BROW, NOSE, RIGHT_EYE, LEFT_EYE, MOUTH]
+        only_test_data -- returns facial vectors for only training data
+        only_test_data -- returns facial vectors for only test data
+        load_cached -- returns image matrix from file if stored, saves processing images again
+
+    :return numpy array [n, 62, 2], where n is number of photos analysed
+    """
+
+    def _create_correct_size_facial_matrix(only_train_data, only_test_data):
+        if only_train_data:
+            facial_featutes_matrix = np.zeros([TRAIN_END_POINT, 68, 2])
+        elif only_test_data:
+            facial_featutes_matrix = np.zeros([PUBLIC_TEST_END_POINT - TRAIN_END_POINT, 68, 2])
+        else:
+            facial_featutes_matrix = np.zeros([PRIVATE_TEST_END_POINT, 68, 2])
+        return facial_featutes_matrix
+
+    def _save_processed_image_matrix(image_matrix, only_train_data, only_test_data):
+        if only_test_data:
+            file_name = TEST_DATA_CACHED_PATH
+        elif only_train_data:
+            file_name = TRAIN_DATA_CACHED_PATH
+        else:
+            file_name = ALL_DATA_CACHED_PATH
+        logger.debug("Saving image_matrix to file {}".format(file_name))
+        np.save(file_name, image_matrix)
+
+    def _load_cached_image_matrix(only_train_data, only_test_data):
+        if only_test_data:
+            file_name = TEST_DATA_CACHED_PATH
+        elif only_train_data:
+            file_name = TRAIN_DATA_CACHED_PATH
+        else:
+            file_name = ALL_DATA_CACHED_PATH
+        logger.debug("Attempting to load cached file: {}".format(file_name))
+        return np.load(file_name)
+
+    if load_cached:
+        try:
+            image_matrix = _load_cached_image_matrix(only_train_data, only_test_data)
+            return image_matrix
+        except FileNotFoundError:
+            logger.debug("No file found, Processing data")
+
+    logger.debug("Getting facial vectors")
+    photos_matrix = _extract_photos_from_file(file_path, False, only_train_data, only_test_data)
+    facial_featutes_matrix = _create_correct_size_facial_matrix(only_train_data, only_test_data)
+    detector = dlib.get_frontal_face_detector()
+    predictor = dlib.shape_predictor(PREDICTOR_PATH)
+    faces_found = 0
+    faces_not_found = 0
+    logger.debug("Starting facial detection and feature extraction")
+    for i in range(len(photos_matrix)):
+        img = photos_matrix[i, :, :]
+        dets = detector(img, 2)
+        if len(dets) != 0:
+            facial_featutes_matrix[i, :, :] = _get_vectors_of_image_from_image(dets, predictor, img)
+            faces_found += 1
+        else:
+            facial_featutes_matrix[i, :, :] = np.zeros([68, 2])
```

Comment:
come to think of it, the `facial_featutes_matrix` is already initialized to zeroes (at initialization), this just overwrites zeroes with a freshly initialized set of zeroes. (unless i'm getting tired and missed something)

to benchmark, i expected significantly worse than this:

```
In [20]: def initializing_each():
    ...:     for i in range(TRAIN_END_POINT):
    ...:         facial_features_matrix[i] = np.zeros([68, 2])
    ...:         

In [21]: def initializing_once():
    ...:     zeroes = np.zeros([68, 2])
    ...:     for i in range(TRAIN_END_POINT):
    ...:         facial_features_matrix[i] = zeroes

In [28]: timeit.timeit('initializing_each()', setup="from __main__ import initializing_each",  number=1000)
Out[28]: 43.244741722010076

In [29]: timeit.timeit('initializing_once()', setup="from __main__ import initializing_once",  number=1000)
Out[29]: 17.67673469998408
```